File: Week3/Code/Vectorize2.R
#Author: Ruth Keane (ruth.keane19@imperial.ac.uk)
#Script: Vectorize2.R
#Desc: vectorizing stochastic Ricker model
#Arguments: none
#Input: none
#Output: none
#Date: Oct 2019
# Runs the stochastic Ricker equation with gaussian fluctuations

rm(list=ls())

stochrick<-function(p0=runif(1000,.5,1.5),r=1.2,K=1,sigma=0.2,numyears=100)
{
  #initialize
  N<-matrix(NA,numyears,length(p0))
  N[1,]<-p0
  
  for (pop in 1:length(p0)){#loop through the populations
    
    for (yr in 2:numyears){ #for each pop, loop through the years

      N[yr,pop] <- N[yr-1,pop] * exp(r * (1 - N[yr - 1,pop] / K) + rnorm(1,0,sigma))
    
    }
  
  }
 return(N)
}

# Now write another function called stochrickvect that vectorizes the above 
# to the extent possible, with improved performance: 
stochrickvect<-function(r=1.2,K=1,sigma=0.2,numyears=100,p0=runif(1000,.5,1.5)){
#make matrix
  N<-matrix(NA,numyears,length(p0))
#add random initial population size
  #pop<-rnorm(1000,0.5,1.5)
  N[1,]<-p0
#loop through the populations
  for (yr in 2:100){ #for each pop, loop through the years
    
    N[yr,] <- N[yr-1,] * exp(r * (1 - N[yr - 1,] / K) + rnorm(1,0,sigma))
  }
 return(N)
}

print("Stochastic Ricker takes:")
print(system.time(res2<-stochrick()))
print("Vectorized Stochastic Ricker takes:")
print(system.time(res2<-stochrickvect()))

File: Week3/Code/Vectorize2.py
#/usr/bin/env python3
#Author: Ruth Keane (ruth.keane19@imperial.ac.uk)
#Script: Vectorize2.py
#Desc: stochastic ricker equation for 1000 populations
#Arguments: none
#Input: none
#Output: none
#Date: Oct 2019
"""stochastic ricker equation for 1000 populations"""
import numpy as np
from timeit import default_timer as timer
from random import uniform
import math
def stochrick(p0=[uniform(0.5, 1.5) for p in range(0, 1000)],r=1.2,K=1,sigma=0.2,numyears=100):
    N=np.zeros([numyears,len(p0)])
    N[0,]=p0
    for pop in range(0,len(p0)):
        for yr in range(1,numyears):
                N[yr,pop] = N[yr-1,pop] * math.exp(r * (1 - N[yr - 1,pop] / K) + np.random.normal(0,sigma,1))        
    return(N)

def stochrickvect(p0=[uniform(0.5, 1.5) for p in range(0, 1000)],r=1.2,K=1,sigma=0.2,numyears=100):
    N=np.zeros([numyears,len(p0)])
    N[0,]=p0
    for yr in range(1,numyears):
        N[yr][:] = N[yr-1][:] * np.exp(r * (1 - N[yr - 1][:] / K) + np.random.normal(0,sigma,1))        
    return(N)

print("Stochastic Ricker takes")
start=timer()
stochrick()
end=timer()
print(end-start)
print("Vectorized Stochastic Ricker takes:")
start=timer()
stochrickvect()
end=timer()
print(end-start)